File: clients/macos/FastSortClientMac/Sources/FastSortClientMac/Services/DanmakuCookieSessionParser.swift
import Foundation

struct DanmakuWeChatSession: Equatable {
    let sessionid: String
    let wxuin: String
}

enum DanmakuCookieSessionParser {
    static func cookieHeader(fromLiveSession liveSession: String?) -> String {
        guard let liveSession else { return "" }
        let raw = liveSession.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !raw.isEmpty else { return "" }
        guard let object = jsonObject(from: raw) else { return raw }
        return cookieHeader(fromJSONObject: object) ?? raw
    }

    static func cookieMap(fromCookieHeader header: String) -> [String: String] {
        header.split(separator: ";").reduce(into: [String: String]()) { result, part in
            let pieces = part.split(separator: "=", maxSplits: 1)
            guard pieces.count == 2 else { return }
            let name = pieces[0].trimmingCharacters(in: .whitespacesAndNewlines)
            let value = pieces[1].trimmingCharacters(in: .whitespacesAndNewlines)
            if !name.isEmpty {
                result[name] = value
            }
        }
    }

    static func wechatSession(fromLiveSession liveSession: String?) -> DanmakuWeChatSession? {
        let object = liveSession.flatMap(jsonObject(from:))
        let cookies = cookieMap(fromCookieHeader: cookieHeader(fromLiveSession: liveSession))
        let sessionid = decodeWeChatValue(
            "\(object?["sessionid"] ?? object?["session_id"] ?? object?["sessionId"] ?? cookies["sessionid"] ?? "")"
        )
        let wxuin = decodeWeChatValue(
            "\(object?["wxuin"] ?? object?["wxUin"] ?? object?["wx_uin"] ?? cookies["wxuin"] ?? "")"
        )
        guard !sessionid.isEmpty, !wxuin.isEmpty else { return nil }
        return DanmakuWeChatSession(sessionid: sessionid, wxuin: wxuin)
    }

    private static func cookieHeader(fromJSONObject object: [String: Any]) -> String? {
        for key in ["cookie", "cookies", "cookieHeader", "cookie_header", "liveSession", "session"] {
            if let value = object[key] as? String,
               !value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                return value
            }
        }
        for key in ["cookies", "cookieMap", "cookie", "session"] {
            if let cookies = object[key] as? [String: Any] {
                return cookieHeader(fromCookieMap: cookies)
            }
            if let items = object[key] as? [[String: Any]] {
                return cookieHeader(fromCookieItems: items)
            }
        }
        return nil
    }

    private static func jsonObject(from raw: String) -> [String: Any]? {
        guard let data = raw.data(using: .utf8) else { return nil }
        return try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    }

    private static func cookieHeader(fromCookieMap cookies: [String: Any]) -> String {
        cookies
            .compactMap { key, value -> String? in
                let text = "\(value)".trimmingCharacters(in: .whitespacesAndNewlines)
                return text.isEmpty || text == "<null>" ? nil : "\(key)=\(text)"
            }
            .sorted()
            .joined(separator: "; ")
    }

    private static func cookieHeader(fromCookieItems items: [[String: Any]]) -> String {
        items
            .compactMap { item -> String? in
                let name = "\(item["name"] ?? "")".trimmingCharacters(in: .whitespacesAndNewlines)
                let value = "\(item["value"] ?? "")".trimmingCharacters(in: .whitespacesAndNewlines)
                return name.isEmpty || value.isEmpty ? nil : "\(name)=\(value)"
            }
            .joined(separator: "; ")
    }

    private static func decodeWeChatValue(_ value: String) -> String {
        var output = value
        for _ in 0..<2 {
            guard output.contains("%25"),
                  let decoded = output.removingPercentEncoding
            else { break }
            output = decoded
        }
        return output
    }
}

File: clients/windows/FastSort.Client.Windows/Core/Danmaku/Cookie/DanmakuCookieSessionParser.cs
using System.Text.Json;

namespace FastSort.Client.Windows.Core.Danmaku.Cookie;

public sealed record DanmakuWeChatSession(string SessionId, string WxUin);

public static class DanmakuCookieSessionParser
{
    private static readonly string[] CookieStringKeys =
    [
        "cookie",
        "cookies",
        "cookieHeader",
        "cookie_header",
        "liveSession",
        "session"
    ];

    private static readonly string[] CookieObjectKeys =
    [
        "cookies",
        "cookieMap",
        "cookie",
        "session"
    ];

    public static string CookieHeaderFromLiveSession(string? liveSession)
    {
        var raw = liveSession?.Trim() ?? "";
        if (string.IsNullOrWhiteSpace(raw))
        {
            return "";
        }

        if (!TryParseJsonObject(raw, out var root))
        {
            return raw;
        }

        return CookieHeaderFromJsonObject(root) is { Length: > 0 } header ? header : raw;
    }

    public static IReadOnlyDictionary<string, string> CookieMapFromCookieHeader(string cookieHeader)
    {
        return cookieHeader
            .Split(';', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries)
            .Select(part => part.Split('=', 2, StringSplitOptions.TrimEntries))
            .Where(parts => parts.Length == 2 && !string.IsNullOrWhiteSpace(parts[0]))
            .GroupBy(parts => parts[0], StringComparer.OrdinalIgnoreCase)
            .ToDictionary(group => group.Key, group => group.Last()[1], StringComparer.OrdinalIgnoreCase);
    }

    public static DanmakuWeChatSession? WeChatSessionFromLiveSession(string? liveSession)
    {
        var objectValues = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        if (!string.IsNullOrWhiteSpace(liveSession) && TryParseJsonObject(liveSession, out var root))
        {
            foreach (var property in root.EnumerateObject())
            {
                if (property.Value.ValueKind is JsonValueKind.String or JsonValueKind.Number)
                {
                    objectValues[property.Name] = property.Value.ToString();
                }
            }
        }

        var cookies = CookieMapFromCookieHeader(CookieHeaderFromLiveSession(liveSession));
        var sessionId = DecodeWeChatValue(GetFirstValue(objectValues, cookies, "sessionid", "session_id", "sessionId"));
        var wxUin = DecodeWeChatValue(GetFirstValue(objectValues, cookies, "wxuin", "wxUin", "wx_uin"));

        return string.IsNullOrWhiteSpace(sessionId) || string.IsNullOrWhiteSpace(wxUin)
            ? null
            : new DanmakuWeChatSession(sessionId, wxUin);
    }

    private static string? CookieHeaderFromJsonObject(JsonElement root)
    {
        foreach (var key in CookieStringKeys)
        {
            if (root.TryGetProperty(key, out var value) &&
                value.ValueKind == JsonValueKind.String &&
                !string.IsNullOrWhiteSpace(value.GetString()))
            {
                return value.GetString();
            }
        }

        foreach (var key in CookieObjectKeys)
        {
            if (!root.TryGetProperty(key, out var value))
            {
                continue;
            }

            if (value.ValueKind == JsonValueKind.Object)
            {
                return CookieHeaderFromCookieMap(value);
            }

            if (value.ValueKind == JsonValueKind.Array)
            {
                return CookieHeaderFromCookieItems(value);
            }
        }

        return null;
    }

    private static string CookieHeaderFromCookieMap(JsonElement map)
    {
        return string.Join("; ", map.EnumerateObject()
            .Select(property =>
            {
                var value = property.Value.ValueKind == JsonValueKind.String
                    ? property.Value.GetString()
                    : property.Value.ToString();
                return new KeyValuePair<string, string>(property.Name, value?.Trim() ?? "");
            })
            .Where(pair => !string.IsNullOrWhiteSpace(pair.Key) &&
                           !string.IsNullOrWhiteSpace(pair.Value) &&
                           !string.Equals(pair.Value, "<null>", StringComparison.OrdinalIgnoreCase))
            .OrderBy(pair => pair.Key, StringComparer.OrdinalIgnoreCase)
            .Select(pair => $"{pair.Key}={pair.Value}"));
    }

    private static string CookieHeaderFromCookieItems(JsonElement items)
    {
        var parts = new List<string>();
        foreach (var item in items.EnumerateArray())
        {
            if (item.ValueKind != JsonValueKind.Object)
            {
                continue;
            }

            var name = TryGetString(item, "name");
            var value = TryGetString(item, "value");
            if (!string.IsNullOrWhiteSpace(name) && !string.IsNullOrWhiteSpace(value))
            {
                parts.Add($"{name}={value}");
            }
        }

        return string.Join("; ", parts);
    }

    private static bool TryParseJsonObject(string raw, out JsonElement root)
    {
        root = default;
        try
        {
            using var document = JsonDocument.Parse(raw);
            if (document.RootElement.ValueKind != JsonValueKind.Object)
            {
                return false;
            }

            root = document.RootElement.Clone();
            return true;
        }
        catch (JsonException)
        {
            return false;
        }
    }

    private static string TryGetString(JsonElement item, string propertyName)
    {
        if (!item.TryGetProperty(propertyName, out var value))
        {
            return "";
        }

        return value.ValueKind == JsonValueKind.String ? value.GetString()?.Trim() ?? "" : value.ToString().Trim();
    }

    private static string GetFirstValue(
        IReadOnlyDictionary<string, string> objectValues,
        IReadOnlyDictionary<string, string> cookieValues,
        params string[] keys)
    {
        foreach (var key in keys)
        {
            if (objectValues.TryGetValue(key, out var objectValue) && !string.IsNullOrWhiteSpace(objectValue))
            {
                return objectValue;
            }

            if (cookieValues.TryGetValue(key, out var cookieValue) && !string.IsNullOrWhiteSpace(cookieValue))
            {
                return cookieValue;
            }
        }

        return "";
    }

    private static string DecodeWeChatValue(string value)
    {
        var output = value;
        for (var i = 0; i < 2; i++)
        {
            if (!output.Contains("%25", StringComparison.OrdinalIgnoreCase))
            {
                break;
            }

            var decoded = Uri.UnescapeDataString(output);
            if (string.Equals(decoded, output, StringComparison.Ordinal))
            {
                break;
            }

            output = decoded;
        }

        return output;
    }
}
